### PR TITLE
Implement "engineering" base-conversion functions

### PR DESCRIPTION
**Pull request by ckp95**
These are: `BIN2OCT`, `BIN2DEC`, `BIN2HEX`, `OCT2BIN`, `OCT2DEC`, `OCT2HEX`, `DEC2BIN`, `DEC2OCT`, `DEC2HEX`, `HEX2BIN`, `HEX2OCT`, `HEX2DEC`. Called "engineering" functions for some reason by the [Office docs](https://support.microsoft.com/en-us/office/engineering-functions-reference-9e99a6b0-1f32-4b25-be6c-c1d1c395e7de).

In hindsight I probably couldn't have picked a more awkward family of functions to implement. They all have weird inconsistencies with each other, they don't handle type conversions the way other functions do (e.g. booleans aren't converted to ints), there are subtle rules regarding what kinds of errors they produce, and they behave differently depending on whether the `places` argument is unspecified vs points to a blank cell. There may be more things I'm forgetting.

However, all that thorniness forced me to develop better testing methodologies. This PR also brings three improvements to the test suite that can be extended elsewhere in the project: easy parametrized testing, auto-generation of tests from .xlsx files, and property-based oracle testing.

## Easy parametrized testing

This isn't a huge deal, it just removes some of the pain points involved in writing parametrized tests. I factored it out as a separate library, you can read more about it [here](https://github.com/ckp95/pytest-parametrize-cases). The upshot is that you can declare lots of test cases in a terse yet readable way.

## Auto-generation of tests from .xlsx files

Instead of laboriously writing individual `xlfunctions_vs_excel` tests, you can now just load an Excel file and automatically generate a labelled test for every single cell that contains a formula. The actual Python code needed to do this is just:

```python
from ..testing import workbook_test_cases, assert_equivalent


@workbook_test_cases("BASES.xlsx")
def test_bases(sheet_value, calculated_value):
    assert_equivalent(result=calculated_value, expected=sheet_value)
```

A test fails if the saved value for that cell is different from how `xlcalculator` evaluates the formula. The test label is the cell address, visible when you use `pytest --verbose`. The `BASES.xlsx` file contains my experiments with the base conversion functions. It produces an ... extravagant number of tests (I wanted to be thorough), many of which are redundant wrt the other test files, so if you want I can strip it down to something more minimal.

## Property-based oracle testing

The [`hypothesis`](https://hypothesis.readthedocs.io/en/latest/) library lets us do property-based testing: assert that certain properties are true of a function or class over domain of input. Basically you define the domain of values on which the function is valid and it randomly explores the input space until it triggers an exception, then reports the failure. The property we are trying to test for is simple: that a Python function replicates the behavior of its Excel counterpart.

To make this happen, we need to be able to take an arbitrary Python value `x`, and an Excel formula, and query Excel for the result of that formula applied to `x`. We can use [`xlwings`](https://www.xlwings.org/) for that. It lets us spawn a live Excel process and drive it from Python, scripting it similarly to how VBA works.

So, I wrote some somewhat hacky code to glue together these things and get a suite of tests that treats a running Excel program as an "oracle" for the example data that hypothesis generates. The end result is you get "declarative" tests that look like this:

```python
@pytest.parametrize_cases(
    ...
    Case(
        "hex2oct",
        formula=HEX2OCT,
        values=given(number=one_of(xl_numbers(), hex_numbers_as_strings(1, 11))),
    ),
    Case(
        "hex2oct-places",
        formula=HEX2OCT,
        values=given(
            number=one_of(xl_numbers(), hex_numbers_as_strings(1, 11)),
            places=xl_numbers(-2, 12),
        ),
    ),
    Case(
        "hex2dec",
        formula=HEX2DEC,
        values=given(number=one_of(xl_numbers(), hex_numbers_as_strings(1, 11))),
    ),
)
def test_against_xlwings(excel_workbook, formula, values):
    fuzz_scalars(excel_workbook, formula, values)
```

This ended up surfacing a whole lot of bugs. I copied the failing examples and put them in the parametrized tests so they'd always get run.

Now, obviously we don't want the test suite reliant on Excel being installed in the environment. I set things up so that the xlwings-reliant tests are skipped if xlwings cannot be imported. So it isn't something that is run in CI; it should be thought of as a development aid to help come up with test cases in a principled way.

I also wrote some property tests that are not reliant on xlwings; they assert that complementary base-conversion functions are inverses of each other (i.e. `x == DEC2BIN(BIN2DEC(x))` for any binary string `x`).

---

I hope it isn't too difficult to follow. I've tried to organize the commits in a logical way. Obviously the way I've presented it here isn't the order in which I developed it; it was a process of trial and error and revision. I hope the improvements to testing can be useful elsewhere in the project too.

Oh and tox shows pytest failing when I run it locally, even though when I run just pytest on its own there are no failures. I can't figure out why. But the error messages I get from tox are the same as those in the most recent CI report on master, which also has a failure. So I'm assuming it isn't something I broke.

### CODE REVIEW

**Comment by strichter** [comment]
Now, I am not the project owner, but I lead a sizable engineering team and the amount of nights that I did not get enough or any sleep at all because of `hypothesis`-based tests is numerous. Builds needs to be stable and predictable. If a random test fails in some arbitrary code that you have no knowledge of, then it is hard to fix and you just wasted a build cycle (which in my company is about 30 minutes.)

Thus I would never allow any tests with variability in my code base.

**Comment by strichter** [comment]
Also, I do not like to rely overly on decorators in tests. They drastically decrease maintainability. In tests it is okay to repeat yourself. It makes them easier to read and maintain. Just write one test function for each test case and describe in the function name what the test is specifically testing.

**Comment by strichter** [comment]
Finally, as far as I know xlwings in Windows specific. This library does not require Windows and having tests that rely on it, even optionally, is no good. (I, for one, do not have Windows and can now not contribute to this project anymore because I cannot adjust and fix xlwings-based tests.)

**Comment by strichter** [comment]
The entire test rig is way out of proportion compared to the handful of functions you wrote.

File: tests/testing.py [hunk 1]
```diff
@@ -1,9 +1,16 @@
...
 from dataclasses import dataclass
 from decimal import Decimal, ROUND_UP, ROUND_DOWN
+from typing import Any, Optional, Callable
+from _pytest.mark.structures import MarkDecorator
+import pytest
 
-from xlcalculator import model, evaluator
+from xlcalculator import xlerrors
+from xlcalculator.model import ModelCompiler
+from xlcalculator.evaluator import Evaluator
```
Comment (strichter) [inline tests/testing.py:13]: Why did you change the imports? I find importing modules much more readbale.

**Comment by bradbase** [comment]
Hi @ckp95 

Thank you for your efforts.

I don't have the brain-power to fully digest all of this PR at the moment. But there are a few things I might be able to offer...

An easy one first.. tox is likely to show pytests failing as there's been a version shift in some underlying libraries which I'm yet to fix. It's easy but the thing that's stopping my brain from internalizing this PR is the same thing that's slowed my progress on drafting, committing and pushing a fix. There is a subsequent bug that gets exposed once the dependency versions are fixed which is getting in the way.

Although @strichter doesn't own the project I have found his advice to be incredibly high quality. He has had an immense input into this library and knows it inside-out/upside down.

I've not had any experience with hypothesis-based testing but the code experience I've had cements the idea that I don't want to be fighting tests as well as code. When it comes to code testing I'm in favor of things that are concretely deterministic. All that said, it doesn't mean there's no space for other kinds of testing suites to put this library through its paces. I absolutely support another code project doing interesting tests with this library. Performance testing is one example. Maybe there's room for an hypothesis-based testing project. The more exercise this code gets the happier I am.

I support copy+paste in tests. Maybe I over-use it but tests need to be very clear. I've tried to short-cut copy+paste in this space before and it's only ever caused more problems than ever solved. Sadly @strichter has been on the receiving end of some of my daft experimentation.

@strichter; xlwings is Windows and Mac. But even then your point is still valid.

I am a _HUGE_ fan of xlwings. I absolutely support maximizing opportunity to integrate xlcalculator and xlwings. But I don't think xlwings has a place in xlcalculator. Again, I support the idea of another project doing the value-add (eg; FlyingKoala).

**Comment by ckp95** [comment]
>Finally, as far as I know xlwings in Windows specific. This library does not require Windows and having tests that rely on it, even optionally, is no good. (I, for one, do not have Windows and can now not contribute to this project anymore because I cannot adjust and fix xlwings-based tests.)

> Now, I am not the project owner, but I lead a sizable engineering team and the amount of nights that I did not get enough or any sleep at all because of `hypothesis`-based tests is numerous. Builds needs to be stable and predictable. If a random test fails in some arbitrary code that you have no knowledge of, then it is hard to fix and you just wasted a build cycle (which in my company is about 30 minutes.)
> 
> Thus I would never allow any tests with variability in my code base.

This is a fair point. I used hypothesis+xlwings as a tool to help me find test cases. Where it found a failing example, I turned that example into its own "real" test case in the other test module. So, strictly speaking the Excel oracle tests have exhausted their value for these functions in particular and they don't need to be in the test suite.

Of course xlwings should never be a dependency of xlcalculator; that would defeat the whole purpose. However I think the approach does still have value as an optional tool for local development. Perhaps I could pull those xlwings parts out and have them as a separate project for benchmarking?

I still think that the "inverses" tests are useful -- it is an important/natural invariant for these functions.

>Why did you change the imports? I find importing modules much more readbale.

Because in the `workbook_test_cases` function I create a variable named `evaluator` that is an instance of the `Evaluator` class, and I wanted to avoid confusion with the `evaluator` module. Similarly with `model`.

**Comment by ckp95** [comment]
>Also, I do not like to rely overly on decorators in tests. They drastically decrease maintainability. In tests it is okay to repeat yourself. It makes them easier to read and maintain. Just write one test function for each test case and describe in the function name what the test is specifically testing.

I resorted to this because there are 12 functions that have overlapping but not identical behaviour and I wanted to avoid repeating myself a ton, and partition the problem space in a manageable way. There's a lot of weird quirks that need to be tested for each of them. They handle errors and type conversions in slightly different ways. A separate test for each one of these behaviors for each function could easily cause the test file to balloon to thousands of lines.

I can go back and do it that way if you want, but it would be difficult to maintain in its own way.

File: xlcalculator/xlfunctions/engineering.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    # if the destination base is dec, we are done
+    if destination == dec:
+        return value
+
+    # magic (handles the 2s-complement-like wrapping behavior):
+    was_negative = value < 0
+    if was_negative:
+        value += 1 << BIT_WIDTHS[destination]
+
+    # otherwise convert to appropriate string representation (return value is (str, bool))
```
Comment (ckp95) [inline xlcalculator/xlfunctions/engineering.py:115]: last part of the comment is wrong

**Comment by strichter** [comment]
@ckp95 I am working on xlcalculator today. If you simplify the test rig, I am happy to merge.

**Comment by strichter** [comment]
I have reimplemented the tests and pushed the functionality of this MR to master.

**Comment by strichter** [comment]
Released in 0.4.0.

**Comment by ckp95** [comment]
Thank you @strichter, I was going to get back to this eventually but you beat me to it.

Also, what is the advantage of the unittest class-style tests vs the pytest function-style tests? i.e. why do this

```python
class EngineeringModuleTest(unittest.TestCase):

    def test_DEC2BIN(self):
        self.assertEqual(
            engineering.DEC2BIN(10), func_xltypes.Text('1010'))

    def test_DEC2BIN_withPlaces(self):
        self.assertEqual(
            engineering.DEC2BIN(10, 5), func_xltypes.Text('01010'))
    ...
```

instead of just

```python
def test_dec2bin:
    assert engineering.DEC2BIN(10) == func_xltypes.Text("1010")

def test_dec2bin_with_places:
    assert engineering.DEC2BIN(10, 5) == func_xltypes.Text("01010")

...
```

I understand that verbosity in a test suite is no vice, but I don't see what the verbosity brings. Not a criticism, just curious if there's some advantage to the class way I don't know about.

**Comment by strichter** [comment]
Because then I can run the tests with other test runners as well. Also the `unittest.TestCase.assert*` functions are well designed. In the functional tests I can simply use `setUp()`/`tearDown()` instead of test fixtures. Classes basically allow you to have state without consequences. When dealing with functions the temptation is too great to store state on the module which can make the test unstable when run multiple times in the same process.
